### ISSUE TITLE
Akka.Persistence does not replay Journal events

### ISSUE THREAD

**Issue by nanderto** [akkadotnet/akka.net#2211]
I am writing an implementation of the Akka.Persistence for Service Fabric, and I don't seem to be able to get the snapshotting working. When it attempts to recover state it gets the latest snapshot but it does not replay the events since the latest snapshot. Its not clear to me If I have simply not wired up the components correctly or if my Implementation of the persistence library is incorrect. 
My actor is a simple counter, my state is just the current count. 
I expect that the Recover<SnapshotOffer> should get called first and then the Recover<Evt> would get called for each journal entry between the last snapshot and the highest sequence number. There is a function ReplayMessagesAsync(...) in the journal that looks like it should do this but it does not get called.
The code for my counter is below, the rest of my code is [:](https://github.com/nanderto/akka.net/tree/ServiceFabric)
`
using Akka.Actor;
using Akka.Persistence;
using Akka.Persistence.ServiceFabric.Journal;
using Akka.Persistence.ServiceFabric.Snapshot;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AkkaPersistence.Actors
{
    public class Counter : ReceivePersistentActor
    {
        public class GetCount { }

        private int counter;

        private CounterState State = new CounterState();

        private int _msgsSinceLastSnapshot = 0;

        public Counter()
        {
            Recover<Evt>(evt =>
            {
                State.Update(evt);
            });

            Recover<SnapshotOffer>(offer => {
                var snapshotEntry = offer.Snapshot as SnapshotEntry;
                if (snapshotEntry != null)
                {
                    State = (CounterState)snapshotEntry.Snapshot;
                }
            });

            Command<string>(str => Persist(str, s =>
            {
                ++counter;
                var evt = new Evt(s);
                State.Update(evt);

                if (++_msgsSinceLastSnapshot % 10 == 0)
                {
                    //time to save a snapshot
                    SaveSnapshot(State.Copy());
                }
            }));

            Command<GetCount>(get => Sender.Tell(State.Count));

            Command<SaveSnapshotSuccess>(success =>
            {
                ServiceEventSource.Current.Message($"Saved snapshot");
                DeleteMessages(success.Metadata.SequenceNr);
            });

            Command<SaveSnapshotFailure>(failure => {
                // handle snapshot save failure...
                ServiceEventSource.Current.Message($"Snapshot failure");
            });

        }

        public override string PersistenceId
        {
            get
            {
                return "counter";
            }
        }
    }

    internal class CounterState
    {
        private long count = 0L;
        
        public long Count
        {
            get { return count; }
            set { count = value; }
        }

        public CounterState(long count)
        {
            this.Count = count;
        }

        public CounterState() : this(0)
        {
        }

        public CounterState Copy()
        {
            return new CounterState(count);
        }

        public void Update(Evt evt)
        {
            ++Count;
        }
    }

    public class Evt
    {
        public Evt(string data)
        {
            Data = data;
        }

        public string Data { get; }

    }

    public class Cmd
    {
        public Cmd(string data)
        {
            Data = data;
        }

        public string Data { get; }
    }

}
`

**Comment by Horusiath** [comment]
1. You're saving `CounterState` as a snapshot (`SaveSnapshot(State.Copy());` call), but when you're recovering from `SnapshotOffer`, you're trying to map snapshot to `SnapshotEntry` instead.
2. You're persisting your events as strings (`Command<string>(str => Persist(str, ...` ), but you're setting recovery handler to `Evt` instead.

**Comment by alexvaluyskiy** [comment]
@nanderto Have you seen this implementation?
https://github.com/akkadotnet/Akka.Persistence.ServiceFabric

**Comment by nanderto** [comment]
Yes, I saw the implementation, I didn't like it when I saw it. The persistence is implemented as a separate service to the actor. I thought it should be imbedded in the actor. That both better matches the other persistence implementations and my view of how it should be implemented. The point of the reliable collection is to allow for state to be saved with the code not separated to a separate database / service layer. If you were building a Service Fabric actor the collection would be imbedded in the actor, so I also wanted to give it parity with SF. So I thought I would give it a go.

**Comment by nanderto** [comment]
@Horusiath Thanks, the second point you made was the trick, I tried swapping from ne implementation to the other and I missed that call. Although so far I can only get it working persisting the string not the Evt.  For some reason when trying to persist the Evt, It never gets from the Persist call to the SaveAsync call.

**Comment by alexvaluyskiy** [comment]
@nanderto That previous implementation uses a very old version of Service Fabric. A lot of things changed there.

**Comment by nanderto** [comment]
Thanks, I have solved my issues. @Horusiath  was correct above, but it took me a coupe of days to realize that my error for 1) above was that I needed to return what was passed in, not my SnapshotEntry which is an implementation detail of my persistence mechanism. 2) above was a simple miss as I translated from saving strings to attempting to save objects as part of the Journal. Finally there was one more issue, that was the underlying issue, and that was, that the serialization was failing. I added a custom serializer (the Wire serializer) for the Journal as well as the already existing SnapshotSerializer and it is now working.
Thanks for the help
